File: apps/web-svelte/src/lib/player/audio-playback-diagnostics.ts
import {
  AUDIO_PLAYBACK_DIAGNOSTIC_EVENT,
  AUDIO_PLAYBACK_PAUSE_SOURCE,
  type AudioPlaybackDiagnosticEventCode,
  type AudioPlaybackPauseSourceCode,
} from "$lib/api/generated/codes";
import type { AudioPlaybackDiagnosticRequest } from "$lib/api/generated/model";

export interface AudioPlaybackDiagnosticSnapshot {
  trackId: string;
  positionSeconds: number;
  durationSeconds: number | null;
  bufferedAheadSeconds: number;
  readyState: number;
  networkState: number;
  paused: boolean;
  ended: boolean;
  playIntent: boolean;
  documentVisible: boolean;
  documentHasFocus: boolean;
  mediaErrorCode: number | null;
}

type DiagnosticSender = (diagnostic: AudioPlaybackDiagnosticRequest) => void;

/**
 * Attributes pause events to the Prismedia action that caused them and measures
 * browser media interruptions from their first signal until active playback resumes.
 */
export class AudioPlaybackDiagnosticReporter {
  private pendingPauseSource: AudioPlaybackPauseSourceCode | null = null;
  private interruptionStartedAt: number | null = null;

  constructor(
    private readonly send: DiagnosticSender,
    private readonly now: () => number = () => performance.now(),
  ) {}

  markPauseSource(source: AudioPlaybackPauseSourceCode): void {
    this.pendingPauseSource = source;
  }

  report(event: AudioPlaybackDiagnosticEventCode, snapshot: AudioPlaybackDiagnosticSnapshot): void {
    let pauseSource: AudioPlaybackPauseSourceCode | null = null;
    let interruptionMilliseconds: number | null = null;

    if (event === AUDIO_PLAYBACK_DIAGNOSTIC_EVENT.pause) {
      pauseSource = this.pendingPauseSource ?? AUDIO_PLAYBACK_PAUSE_SOURCE.browser;
      this.pendingPauseSource = null;
      if (pauseSource === AUDIO_PLAYBACK_PAUSE_SOURCE.browser && snapshot.playIntent) {
        this.interruptionStartedAt ??= this.now();
      } else {
        this.interruptionStartedAt = null;
      }
    } else if (
      snapshot.playIntent &&
      (event === AUDIO_PLAYBACK_DIAGNOSTIC_EVENT.waiting ||
        event === AUDIO_PLAYBACK_DIAGNOSTIC_EVENT.stalled)
    ) {
      this.interruptionStartedAt ??= this.now();
    } else if (event === AUDIO_PLAYBACK_DIAGNOSTIC_EVENT.playing) {
      this.pendingPauseSource = null;
      if (this.interruptionStartedAt !== null) {
        interruptionMilliseconds = Math.max(0, Math.round(this.now() - this.interruptionStartedAt));
        this.interruptionStartedAt = null;
      }
    } else if (event === AUDIO_PLAYBACK_DIAGNOSTIC_EVENT.error) {
      this.pendingPauseSource = null;
      if (snapshot.playIntent) {
        this.interruptionStartedAt ??= this.now();
      } else {
        this.interruptionStartedAt = null;
      }
    }

    this.send({
      event,
      ...snapshot,
      pauseSource,
      interruptionMilliseconds,
    });
  }
}

File: apps/backend/src/Prismedia.Contracts/Playback/AudioPlaybackDiagnostics.cs
using Prismedia.Domain.Entities;

namespace Prismedia.Contracts.Playback;

/// <summary>
/// One low-volume browser audio lifecycle transition used to diagnose intermittent playback stalls.
/// </summary>
/// <param name="Event">Media lifecycle transition being reported.</param>
/// <param name="TrackId">Audio track loaded by the media element.</param>
/// <param name="PositionSeconds">Media playhead position when the transition occurred.</param>
/// <param name="DurationSeconds">Known media duration, when finite.</param>
/// <param name="BufferedAheadSeconds">Contiguous buffered media available after the playhead.</param>
/// <param name="ReadyState">Browser HTMLMediaElement readyState numeric value.</param>
/// <param name="NetworkState">Browser HTMLMediaElement networkState numeric value.</param>
/// <param name="Paused">Whether the media element reports itself paused.</param>
/// <param name="Ended">Whether the media element reached its natural end.</param>
/// <param name="PlayIntent">Whether Prismedia still intends playback to be active.</param>
/// <param name="DocumentVisible">Whether the player document is visible.</param>
/// <param name="DocumentHasFocus">Whether the player document has focus.</param>
/// <param name="PauseSource">Prismedia action owning a pause, or Browser when no action owns it.</param>
/// <param name="InterruptionMilliseconds">Elapsed time from the first waiting signal until playback recovered.</param>
/// <param name="MediaErrorCode">HTMLMediaElement error code, when an error exists.</param>
public sealed record AudioPlaybackDiagnosticRequest(
    AudioPlaybackDiagnosticEvent Event,
    Guid TrackId,
    double PositionSeconds,
    double? DurationSeconds,
    double BufferedAheadSeconds,
    int ReadyState,
    int NetworkState,
    bool Paused,
    bool Ended,
    bool PlayIntent,
    bool DocumentVisible,
    bool DocumentHasFocus,
    AudioPlaybackPauseSource? PauseSource,
    long? InterruptionMilliseconds,
    int? MediaErrorCode);
